utils/hashPassword.js

```diff
@@ -1,16 +1,132 @@
+const PBKDF2_KEY_BITS = PBKDF2_DIGEST_BYTES * 8
+
+/**
+ * Hash a password using Node's asynchronous pbkdf2 (key derivation) function.
+ *
+ * Returns a self-contained buffer which can be arbitrarily encoded for storage
+ * that contains all the data needed to verify a password.
+ *
+ * @param {!String} password
+ * @return {Promise<Buffer>} Generated hash (with encoded salt and iterations)
+ */
+function hashPassword (password, customIterations) {
+  const iterations = (process.env.NODE_ENV !== 'unit')
+    ? customIterations || PBKDF2_ITERATIONS
+    : PBKDF2_ITERATIONS_TEST
+
+  if (typeof password === 'string') {
+    password = new Buffer(password, 'utf8')
+  }
+
+  return new Promise((resolve, reject) => {
+    // Generate a salt for pbkdf2
+    crypto.randomBytes(PBKDF2_SALT_BYTES, (err, salt) => {
+      if (err) {
+        return reject(err)
+      }
+
+      try {
+        crypto.pbkdf2(
+          password,
+          salt,
+          iterations,
+          PBKDF2_KEY_BITS,
+          PBKDF2_DIGEST,
+          (err, hash) => {
+            if (err) {
+              return reject(err)
+            }
+
+            var combined = new Buffer(hash.length + salt.length + 8)
+
+            // Include the size of the salt so that we can, during verification,
+            // figure out how much of the hash is salt
+            combined.writeUInt32BE(salt.length, 0, true)
+
+            // Similarly, include the iteration count
+            combined.writeUInt32BE(PBKDF2_ITERATIONS, 4, true)
+
+            salt.copy(combined, 8)
+            hash.copy(combined, salt.length + 8)
+
+            resolve(combined)
+          }
+        )
+      } catch (err) {
+        reject(err)
       }
-      bcrypt.hash(password, salt, function (err, hash) {
-        return err ? reject(err) : resolve(hash)
-      })
     })
   })
 }
+
+/**
+ * Verify a password using Node's asynchronous pbkdf2 (key derivation) function.
+ *
+ * Accepts a hash and salt generated by hashPassword, and returns whether the
+ * hash matched the password (as a boolean).
+ *
+ * @param {!String} password
+ * @param {!Buffer} combined Buffer containing hash and salt as generated by
+ *   hashPassword.
+ * @param {Promise<Boolean>} Whether the password matched the hash
+ */
+function verifyPassword (password, combined) {
+  // Extract the salt and hash from the combined buffer
+  const saltBytes = combined.readUInt32BE(0)
+  const hashBytes = combined.length - saltBytes - 8
+  const iterations = combined.readUInt32BE(4)
+  const salt = combined.slice(8, saltBytes + 8)
+  const hash = combined.slice(saltBytes + 8)
+
+  return new Promise((resolve, reject) => {
+    // Verify the salt and hash against the password
+    crypto.pbkdf2(password, salt, iterations, hashBytes, PBKDF2_DIGEST, (err, verify) => {
+      if (err) {
+        return reject(err)
+      }
+
+      resolve(Buffer.compare(verify, hash) === 0)
```

Comment:
`resolve(verify.equals(hash))` would be a little simpler. ([Buffer#equals](https://nodejs.org/api/buffer.html#buffer_buf_equals_otherbuffer))